modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/IndexBuilderTxRwOperationTracker.java

```diff
@@ -0,0 +1,203 @@
+
+    private final NavigableMap<Integer, CompletableFuture<Void>> awaitCompleteOperationsFutureByCatalogVersion
+            = new ConcurrentSkipListMap<>();
+
+    private final NavigableMap<Integer, TxRwOperationCounter> operationCounterByCatalogVersion = new ConcurrentSkipListMap<>();
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    private final CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+
+    @Override
+    public void close() {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        closeFuture.completeExceptionally(new NodeStoppingException());
+    }
+
+    /**
+     * Updating the minimum available catalog version on which new RW transaction operations can be started.
+     *
+     * <p>It is expected that it will be updated when the index transitions to the backfield state and will increase monotonically.</p>
+     *
+     * @param catalogVersion Catalog version in which the new index appeared.
+     */
+    void updateMinAllowedCatalogVersionForStartOperation(int catalogVersion) {
+        inBusyLock(busyLock, () -> {
+            minAllowedCatalogVersionForStartOperation.updateAndGet(previousCatalogVersion -> {
+                assert catalogVersion > previousCatalogVersion :
+                        "It should only grow: [previous=" + previousCatalogVersion + ", new=" + catalogVersion + "]";
+
+                return catalogVersion;
+            });
+
+            Collection<CompletableFuture<Void>> futures = awaitCompleteOperationsFutureByCatalogVersion.headMap(catalogVersion, true)
+                    .values();
+
+            futures.forEach(future -> future.complete(null));
+
+            futures.clear();
+        });
+    }
+
+    /**
+     * Waits for RW transactions operations to complete before the requested catalog version.
+     *
+     * <p>It first waits for the requested catalog version to be reached in {@link #updateMinAllowedCatalogVersionForStartOperation(int)}
+     * and then waits for all RW transaction operations to complete.</p>
+     *
+     * @param catalogVersion Catalog version in which the new index appeared.
+     */
+    CompletableFuture<Void> awaitCompleteTxRwOperations(int catalogVersion) {
+        return inBusyLock(busyLock, () -> {
+            // This code is needed to avoid races with updateAllowedCatalogVersionForStartOperation.
+            CompletableFuture<Void> awaitFuture = awaitCompleteOperationsFutureByCatalogVersion.computeIfAbsent(
+                    catalogVersion,
+                    i -> new CompletableFuture<>()
+            );
+
+            if (catalogVersion <= minAllowedCatalogVersionForStartOperation.get()) {
+                awaitCompleteOperationsFutureByCatalogVersion.remove(catalogVersion);
```

Comment:
Also, it seems that the field should be renamed to help understand what futures it has. Like `minAllowedVersionRaiseFutures`.

Reply:
why not